### PR TITLE
Issue #1676 - Move remove AppInstanceLabelKey, ConfigManagementPlugins, ResourceOverrides, ResourceExclusions, ResourceInclusions settings from ArgoCDSettings structure

### LINKED ISSUE #1199: Support hot-reload for resource.exclusions
Currently, when the user changes `resource.exclusions`, we do not stop the existing watches. We should restart our watches when this event occurs.

### LINKED ISSUE #1774: Data race in app controller
**Describe the bug**

Application controller has the following data race:

<details><summary>DATA RACE</summary>
<p>

```
==================
WARNING: DATA RACE
Write at 0x00c00016dc78 by goroutine 72:
  github.com/argoproj/argo-cd/controller.(*ApplicationController).watchSettings()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:987 +0x4f9

Previous read at 0x00c00016dc78 by goroutine 43:
  github.com/argoproj/argo-cd/util/settings.(*ArgoCDSettings).GetAppInstanceLabelKey()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/util/settings/settings.go:892 +0x57
  github.com/argoproj/argo-cd/controller/cache.(*clusterInfo).onNodeUpdated()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cluster.go:469 +0x27c
  github.com/argoproj/argo-cd/controller/cache.(*clusterInfo).processEvent()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cluster.go:458 +0x58a
  github.com/argoproj/argo-cd/controller/cache.(*clusterInfo).watchEvents.func1()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cluster.go:241 +0xb99
  github.com/argoproj/argo-cd/util.RetryUntilSucceed()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/util/util.go:99 +0x2db
  github.com/argoproj/argo-cd/controller/cache.(*clusterInfo).watchEvents()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cluster.go:193 +0x505

Goroutine 72 (running) created at:
  github.com/argoproj/argo-cd/controller.(*ApplicationController).Run()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:267 +0x260

Goroutine 43 (running) created at:
  github.com/argoproj/argo-cd/controller/cache.(*clusterInfo).startMissingWatches()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cluster.go:180 +0x72f
  github.com/argoproj/argo-cd/controller/cache.(*clusterInfo).sync()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cluster.go:306 +0xc4b
  github.com/argoproj/argo-cd/controller/cache.(*clusterInfo).ensureSynced()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cluster.go:325 +0x160
  github.com/argoproj/argo-cd/controller/cache.(*liveStateCache).getSyncedCluster()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cache.go:111 +0x145
  github.com/argoproj/argo-cd/controller/cache.(*liveStateCache).IsNamespaced()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/cache/cache.go:131 +0xad
  github.com/argoproj/argo-cd/controller.DeduplicateTargetObjects()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/state.go:163 +0x241
  github.com/argoproj/argo-cd/controller.(*appStateManager).CompareAppState()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/state.go:262 +0xeb3
  github.com/argoproj/argo-cd/controller.(*ApplicationController).processAppRefreshQueueItem()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:663 +0x1092
  github.com/argoproj/argo-cd/controller.(*ApplicationController).Run.func2()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/controller/appcontroller.go:279 +0x57
  github.com/argoproj/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x7e
  github.com/argoproj/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0x13d
  github.com/argoproj/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
      /Users/amatyushentsev/root/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x96
==================
```

</p>
</details>

Due to this data race controller might miss setting update event and incorrectly reconcile applications

### PR DESCRIPTION
**Pull request by alexmt**
Part I of settings management refactoring. 

Also Closes: https://github.com/argoproj/argo-cd/issues/1774
Closes: https://github.com/argoproj/argo-cd/issues/1199

### CODE REVIEW

File: controller/appcontroller.go [hunk 1]
```diff
@@ -974,45 +967,3 @@ func toggledAutomatedSync(old *appv1.Application, new *appv1.Application) bool {
 	// nothing changed
 	return false
 }
-
-func (ctrl *ApplicationController) watchSettings(ctx context.Context) {
-	updateCh := make(chan *settings_util.ArgoCDSettings, 1)
```
Comment (alexec) [inline controller/appcontroller.go]: Yay!

File: controller/cache/cache.go [hunk 2]
```diff
@@ -47,32 +54,51 @@ func GetTargetObjKey(a *appv1.Application, un *unstructured.Unstructured, isName
...
+	onAppUpdated      AppUpdatedHandler
+	kubectl           kube.Kubectl
+	settingsMgr       *settings.SettingsManager
+	metricsServer     *metrics.MetricsServer
+	cacheSettingsLock *sync.Mutex
+	cacheSettings     *cacheSettings
+}
+
+func (c *liveStateCache) loadCacheSettings() (*cacheSettings, error) {
+	appInstanceLabelKey, err := c.settingsMgr.GetAppInstanceLabelKey()
```
Comment (alexec) [inline controller/cache/cache.go]: No test?

File: controller/cache/cache.go [hunk 3]
```diff
@@ -161,8 +187,55 @@ func isClusterHasApps(apps []interface{}, cluster *appv1.Cluster) bool {
...
 }
 
+func (c *liveStateCache) getCacheSettings() *cacheSettings {
+	c.cacheSettingsLock.Lock()
+	defer c.cacheSettingsLock.Unlock()
+	return c.cacheSettings
+}
+
+func (c *liveStateCache) watchSettings(ctx context.Context) {
+	updateCh := make(chan *settings.ArgoCDSettings, 1)
```
Comment (alexec) [inline controller/cache/cache.go]: No test?

File: controller/cache/cache.go [hunk 4]
```diff
@@ -161,8 +187,55 @@ func isClusterHasApps(apps []interface{}, cluster *appv1.Cluster) bool {
...
+	}
+	log.Info("shutting down settings watch")
+	c.settingsMgr.Unsubscribe(updateCh)
+	close(updateCh)
+}
+
 // Run watches for resource changes annotated with application label on all registered clusters and schedule corresponding app refresh.
-func (c *liveStateCache) Run(ctx context.Context) {
+func (c *liveStateCache) Run(ctx context.Context) error {
+	cacheSettings, err := c.loadCacheSettings()
```
Comment (alexec) [inline controller/cache/cache.go]: No test?

File: server/application/application.go [hunk 5]
```diff
@@ -178,22 +178,28 @@ func (s *Server) GetManifests(ctx context.Context, q *application.ApplicationMan
...
 	}
 	helmRepos, err := s.db.ListHelmRepos(ctx)
 	if err != nil {
 		return nil, err
 	}
-	tools := make([]*appv1.ConfigManagementPlugin, len(settings.ConfigManagementPlugins))
-	for i := range settings.ConfigManagementPlugins {
-		tools[i] = &settings.ConfigManagementPlugins[i]
+
+	plugins, err := s.settingsMgr.GetConfigManagementPlugins()
```
Comment (alexec) [inline server/application/application.go]: No tests?

File: test/e2e/fixture/fixture.go [hunk 6]
```diff
@@ -140,15 +143,40 @@ func Settings(consumer func(s *settings.ArgoCDSettings)) {
...
+			delete(cm.Data, "resource.customizations")
+		}
+		return nil
 	})
 }
 
-func SetConfigManagementPlugin(plugin v1alpha1.ConfigManagementPlugin) {
-	Settings(func(s *settings.ArgoCDSettings) {
-		s.ConfigManagementPlugins = []v1alpha1.ConfigManagementPlugin{plugin}
+func SetConfigManagementPlugin(plugin ...v1alpha1.ConfigManagementPlugin) {
```
Comment (alexec) [inline test/e2e/fixture/fixture.go]: Pluralise name?

File: util/settings/resources_filter.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
+package settings
+
+type ResourcesFilter struct {
```
Comment (alexec) [inline util/settings/resources_filter.go]: Looks good.

File: util/settings/settings.go [hunk 8]
```diff
@@ -175,16 +164,84 @@ func (mgr *SettingsManager) GetSecretsLister() (v1listers.SecretLister, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		rf.ResourceExclusions = excludedResources
+	}
+	return rf, nil
+}
+
+func (mgr *SettingsManager) GetAppInstanceLabelKey() (string, error) {
+	argoCDCM, err := mgr.getConfigMap()
```
Comment (alexec) [inline util/settings/settings.go]: No test?

File: util/settings/settings.go [hunk 9]
```diff
@@ -175,16 +164,84 @@ func (mgr *SettingsManager) GetSecretsLister() (v1listers.SecretLister, error) {
...
+	}
+	label := argoCDCM.Data[settingsApplicationInstanceLabelKey]
+	if label == "" {
+		return common.LabelKeyAppInstance, nil
+	}
+	return label, nil
+}
+
+func (mgr *SettingsManager) GetConfigManagementPlugins() ([]v1alpha1.ConfigManagementPlugin, error) {
+	argoCDCM, err := mgr.getConfigMap()
```
Comment (alexec) [inline util/settings/settings.go]: No test?

File: util/settings/settings.go [hunk 10]
```diff
@@ -175,16 +164,84 @@ func (mgr *SettingsManager) GetSecretsLister() (v1listers.SecretLister, error) {
...
+		if err != nil {
+			return nil, err
+		}
+	}
+	return plugins, nil
+}
+
+// GetResouceOverrides loads Resource Overrides from argocd-cm ConfigMap
+func (mgr *SettingsManager) GetResourceOverrides() (map[string]v1alpha1.ResourceOverride, error) {
+	argoCDCM, err := mgr.getConfigMap()
```
Comment (alexec) [inline util/settings/settings.go]: No test?

**Comment by alexec** [comment]
It seems a lot of test coverage is missing from this PR?

File: server/application/application.go (hunk 5 above)
Comment (alexmt) [reply server/application/application.go]: added tests

File: test/e2e/fixture/fixture.go (hunk 6 above)
Comment (alexmt) [reply test/e2e/fixture/fixture.go]: done

File: util/settings/settings.go (hunk 8 above)
Comment (alexmt) [reply util/settings/settings.go]: added

File: util/settings/settings.go (hunk 9 above)
Comment (alexmt) [reply util/settings/settings.go]: added tests

File: util/settings/settings.go (hunk 10 above)
Comment (alexmt) [reply util/settings/settings.go]: added tests

**Review by alexec — APPROVED** [review]

**Pull request merged**